### ISSUE TITLE
`AsyncCommands::scan` is broken

### ISSUE THREAD

**Issue by emchristiansen** [redis-rs/redis-rs#759]
This small modification of [this example](https://github.com/redis-rs/redis-rs/blob/f7409d0f86e11293d924e8f39eb4119a9f41add6/redis/examples/async-scan.rs) fails:

```rust
let client = redis::Client::open("redis://127.0.0.1/0").unwrap();
let mut con = client.get_async_connection().await?;

for i in 0..100 {
  con.set(format!("async-key{i}"), b"foo").await?;
}

let iter: AsyncIter<String> = con.scan().await?;
let mut keys: Vec<_> = iter.collect().await;
keys.sort();

for k in &keys {
  println!("{k}");
}
assert_eq!(keys.len(), 100);

Ok(())
```

When I run it multiple times I get different numbers for the size of `keys`: 632, 790, 472, 695, ...

BTW, I started looking into this because `scan` sometimes doesn't return all the keys in a DB.
For me, it was returning only 18 keys for a DB with 3.5M keys, and then hanging indefinitely.

**Comment by djc** [comment]
Ugh, that doesn't sound good. Would you be able to investigate more?

**Comment by emchristiansen** [comment]
I'm sorry, but I doubt I'll have time to look into it anytime soon.

**Comment by PhotonQuantum** [comment]
I think there are two problems involved:
1. Errors in cursor iteration are silently ignored and end the stream. https://github.com/redis-rs/redis-rs/blob/7a9802132f1e544bd60483e8242a6db3694e658f/redis/src/cmd.rs#L116
2. The `Stream` implementation is broken, repeatedly aborting pending futures, flooding the server with duplicate commands, and causing a block. https://github.com/redis-rs/redis-rs/issues/537

IMO the current implementation of AsyncIter is seriously broken and needs to be reworked as soon as possible.

**Comment by djc** [comment]
@PhotonQuantum would you be able to help rework `AsyncIter`? I think I probably agree but unfortunately I don't have much time to look into these issues in more detail right now.

**Comment by jaymell** [comment]
It's worth noting that we have a PR already that fixes this issue -- #597. It may not be the most elegant solution but doesn't break any APIs and fixes the problem we're seeing here. I suggest we reconsider merging that PR. This is a long-standing serious issue that deserves a fix.

**Comment by PhotonQuantum** [comment]
#597 fixes the second problem I've mentioned. Regarding to the first problem, the fix itself is fairly straightforward, but changing the output to Result is a breaking change. I'm not sure how to make the transition as smooth as possible.

**Comment by jaymell** [comment]
@PhotonQuantum Thanks for the clarification there. It's definitely worth considering breaking changes here in order to better handle iteration failures, though that probably merits a separate issue that can be handled separately from the fundamental brokenness of `AsyncIter` we see in the repro above.

**Comment by IndigoCurnick** [comment]
I think I'm running into this issue. Currently using redis 0.25.3 and Rust 1.78.0
I made two implementations that would get all keys in cluster

```
async fn scan_space(cluster_connection: &Arc<RwLock<ClusterConnection>>) -> Vec<String> {
    let mut redis_scanner = cluster_connection.write().await;
    let mut keys = vec![];
    let mut scan: AsyncIter<String> = redis_scanner.scan().await.unwrap();

    if let Some(key) = scan.next_item().await {
        keys.push(key);
    }

    return keys;
}

async fn key_space(cluster_connection: &Arc<RwLock<ClusterConnection>>) -> Vec<String> {
    let mut redis_scanner = cluster_connection.write().await;

    let keys = redis_scanner.keys("*").await.unwrap();

    return keys;
}
```

And then this little dummy test script 

```
    let init_scan = scan_space(redis_connection).await;
    let init_keys = key_space(redis_connection).await;
    info!("Initial keys from scan: {:?}", init_scan);
    info!("Initial keys from keys: {:?}", init_keys);

    let mut redis_writer = redis_connection.write().await;

    for i in 0..10 {
        let res: Result<String, RedisError> = redis_writer.set(format!("key-{}", i), i).await;

        match res {
            Ok(x) => info!("Successfully inserted {} with code {}", i, x),
            Err(y) => panic!("{}", y),
        };
    }

    drop(redis_writer);

    let post_insert_scan = scan_space(redis_connection).await;
    let post_insert_keys = key_space(redis_connection).await;

    info!("Post insert keys from scan: {:?}", post_insert_scan);
    info!("Post insert keys from keys: {:?}", post_insert_keys);

    let mut redis_deleter = redis_connection.write().await;
    for key in post_insert_scan {
        info!("Deleting key {}", key);
        let res: () = redis_deleter.del(key).await.unwrap();
    }

    drop(redis_deleter);

    let post_delete_scan = scan_space(redis_connection).await;
    let post_delete_keys = key_space(redis_connection).await;

    info!("Post delete keys from scan: {:?}", post_delete_scan);
    info!("Post delete keys from keys: {:?}", post_delete_keys);
```

Each time I run this the keys version always gets all of the keys perfectly, but the scan version is usually only able to get about 2-4 of the keys. 

Looks like a fix for this was merged in over a year ago, but this issue is still open. Any plans to release this fix soon?

**Comment by nihohit** [comment]
This is probably not a client issue - SCAN is ill-defined for clusters, see the discussion here - https://github.com/valkey-io/valkey/issues/33

Adding support for cluster scan requires checking which nodes were scanned through, and whether the cluster experienced a topology change.

**Comment by IndigoCurnick** [comment]
I see - in that thread they discuss sending a SCAN command to each node of the cluster. Is that possible with the `ClusterConnection`?

**Comment by nihohit** [comment]
yes, see discussion here - https://github.com/redis-rs/redis-rs/issues/1230

**Comment by IndigoCurnick** [comment]
Thanks!

Are there any examples of how one would implement your suggestion in #1230 - I'm having trouble figuring out how I should actually do it

Also, I found [https://github.com/amazon-contributing/redis-rs/pull/141](https://github.com/amazon-contributing/redis-rs/pull/141) which I think is a feature that would come in really handy. Any indication of if/when that might be release in the official crate?
